### PR TITLE
Fix #161: Exploration player contentcard supports rich-text

### LINKED ISSUE #161: UI Structure: ContentCardView
This is tracking implementing the initial structure for the content card view that's being fully implemented in #26. This doesn't need to support rich-text support yet, but at least having a way to process Oppia's custom HTML tags seems worthwhile since it may be complex and require infrastructural work.

This also doesn't need to have any special data associated with it (hence why it has no domain controller blockers).

### PR DESCRIPTION
**Pull request by veena14cs**
## Explanation

This PR is replica of [#125](https://github.com/oppia/oppia-android/pull/125). It includes rich-text component that parses HTML content. This PR also extracts image from the HTML content.

### CODE REVIEW

File: app/src/main/java/org/oppia/app/fragment/FragmentComponent.kt [hunk 1]
```diff
@@ -24,4 +25,5 @@ interface FragmentComponent {
   fun inject(homeFragment: HomeFragment)
   fun inject(stateFragment: StateFragment)
   fun inject(conceptCardFragment: ConceptCardFragment)
+  fun inject(conceptCardFragment: ContentListFragment)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/fragment/FragmentComponent.kt]: Check this variable name, it should be `contentListFragment`

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import androidx.databinding.library.baseAdapters.BR
+import androidx.recyclerview.widget.RecyclerView
+import kotlinx.android.synthetic.main.content_card_items.view.*
+import org.oppia.app.R
+import org.oppia.app.databinding.ContentCardItemsBinding
+import org.oppia.app.databinding.LearnersCardItemBinding
+import org.oppia.data.backends.gae.model.GaeSubtitledHtml
+import org.oppia.util.data.UrlImageParser
+
+/** Adapter to bind the contents to the [RecyclerView].It handles rich-text content  */
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Update this comment to 
`/** Adapter to bind the contents to the [RecyclerView]. It handles rich-text content.  */`

Also, do check what differences are there in the two comments.

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+class ContentCardAdapter(
+  private val context: Context,
+  val contentList: MutableList<GaeSubtitledHtml>
+) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
+
+  private val VIEW_TYPE_CONTENT = 1
+  private val VIEW_TYPE_LEARNER = 2
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Should we name this as `VIEW_TYPE_INTERACTION` ?

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        val binding =
+          DataBindingUtil.inflate<LearnersCardItemBinding>(inflater, R.layout.learners_card_item, parent, false)
+        LearnersViewHolder(binding)
+      }
+      else -> throw IllegalArgumentException("Invalid view type")
+    }
+  }
+
+  override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Remove this blank line.

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    when (holder.itemViewType) {
+      VIEW_TYPE_CONTENT -> (holder as ContentViewHolder).bind(contentList!!.get(position).html)
+      VIEW_TYPE_LEARNER -> (holder as LearnersViewHolder).bind(contentList!!.get(position).html)
+    }
+  }
+
+  // Determines the appropriate ViewType according to the sender of the message.
+  override fun getItemViewType(position: Int): Int {
+
+    return if (!contentList!!.get(position).contentId!!.contains("content") && !contentList!!.get(position).contentId!!.contains(
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Break this condition in multiple lines, mostly we don't want more than 80 characters in single line.

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  private fun parseHtml(rawString: String?, tvContents: TextView): Spannable {
+    val html: Spannable
+    var htmlContent = rawString
+
+    var CUSTOM_TAG = "oppia-noninteractive-image"
+    var HTML_TAG = "img"
+    var CUSTOM_ATTRIBUTE = "filepath-with-value"
+    var HTML_ATTRIBUTE = "src"
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: All these can be declared at the top of class as `const val`

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+    if (htmlContent!!.contains(CUSTOM_TAG)) {
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Remove this extra line

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+  inner class LearnersViewHolder(val binding: ViewDataBinding) : RecyclerView.ViewHolder(binding.root) {
+
+    fun bind(rawString: String?) {
+
+      binding.setVariable(BR.htmlContent, rawString)
+      binding.executePendingBindings();
+
+      val html: Spannable = parseHtml(rawString, binding.root.tvContents)
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Remove this extra line

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    } else {
+      html = Html.fromHtml(htmlContent, imageGetter, null) as Spannable
+    }
+    return html
+  }
+
+  inner class LearnersViewHolder(val binding: ViewDataBinding) : RecyclerView.ViewHolder(binding.root) {
+
+    fun bind(rawString: String?) {
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Remove this extra line

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  @ApplicationContext private val context: Context,
+  private val fragment: Fragment
+) {
+
+  private lateinit var binding: ContentListFragmentBinding
+  var contentCardAdapter: ContentCardAdapter? = null
+  var contentList: MutableList<GaeSubtitledHtml> = ArrayList()
+
+  fun handleCreateView(inflater: LayoutInflater, container: ViewGroup?): View? {
+
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt:23]: Remove this extra line

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+      binding.recyclerView.layoutManager = LinearLayoutManager(context)
+      contentCardAdapter = ContentCardAdapter(context, contentList);
+      binding.contentCardAdapter = ContentCardAdapter(context, contentList);
+    }
+    fetchDummyExplorations()
+
+    return binding.root
+  }
+
+  //TODO (#121) :Replace this once interface for ExplorationDataController is available
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt]: Update comment:

// TODO (#121) :Replace this once interface for ExplorationDataController is available.

File: app/src/main/res/drawable/bg_blue_card.xml [hunk 12]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+       android:padding="16dp"
+       android:shape="rectangle">
+  <stroke
+    android:width="1.5dp"
```
Comment (rt4914) [inline app/src/main/res/drawable/bg_blue_card.xml]: `1.5 dp` why? I suggest not using such values,  probably we should be them in multiple of `4` or sometimes `2` if required.

File: app/src/main/res/drawable/bg_blue_card.xml [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+    android:color="@color/blue_200"/>
+  <corners android:radius="10dp"/>
```
Comment (rt4914) [inline app/src/main/res/drawable/bg_blue_card.xml]: Either `8dp` or `10dp`

File: app/src/main/res/drawable/bg_white_card.xml [hunk 14]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+       android:padding="16dp"
+       android:shape="rectangle">
+  <stroke
+    android:width="1.5dp"
```
Comment (rt4914) [inline app/src/main/res/drawable/bg_white_card.xml]: Ditto

File: app/src/main/res/drawable/bg_white_card.xml [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+    android:color="@color/black"/>
+  <corners android:radius="10dp"/>
```
Comment (rt4914) [inline app/src/main/res/drawable/bg_white_card.xml]: Ditto

File: app/src/main/res/layout/content_card_items.xml [hunk 16]
```diff
@@ -0,0 +1,22 @@
+<layout xmlns:android="http://schemas.android.com/apk/res/android">
+  <data>
+    <variable
+      name="htmlContent"
+      type="String"/>
+  </data>
+  <LinearLayout
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:layout_margin="5dp"
```
Comment (rt4914) [inline app/src/main/res/layout/content_card_items.xml]: Ditto for `dp`

File: app/src/main/res/layout/content_card_items.xml [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+    android:background="@drawable/bg_blue_card"
+    android:orientation="vertical">
+    <TextView
+      android:id="@+id/tvContents"
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:layout_margin="5dp"
```
Comment (rt4914) [inline app/src/main/res/layout/content_card_items.xml]: Ditto for `dp`

File: app/src/main/res/layout/content_list_fragment.xml [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    tools:context=".player.content.ContentListFragment">
+    <androidx.recyclerview.widget.RecyclerView
+      android:id="@+id/recyclerView"
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:layout_alignParentTop="true"
+      android:layout_marginStart="8dp"
+      android:layout_marginTop="4dp" android:layout_marginEnd="8dp"
```
Comment (rt4914) [inline app/src/main/res/layout/content_list_fragment.xml]: Indentation issue

File: app/src/main/res/layout/learners_card_item.xml [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+      name="htmlContent"
+      type="String"/>
+  </data>
+  <RelativeLayout
+    android:layout_width="wrap_content"
+    android:layout_height="wrap_content"
+    android:layout_marginLeft="22dp"
+    android:layout_marginTop="5dp"
+    android:layout_marginRight="5dp"
+    android:layout_marginBottom="5dp"
```
Comment (rt4914) [inline app/src/main/res/layout/learners_card_item.xml]: Ditto for dp

File: app/src/main/res/layout/learners_card_item.xml [hunk 20]
```diff
@@ -0,0 +1,25 @@
+<layout xmlns:android="http://schemas.android.com/apk/res/android">
+  <data>
+    <variable
+      name="htmlContent"
+      type="String"/>
+  </data>
+  <RelativeLayout
+    android:layout_width="wrap_content"
+    android:layout_height="wrap_content"
+    android:layout_marginLeft="22dp"
```
Comment (rt4914) [inline app/src/main/res/layout/learners_card_item.xml]: Either `20` or `24`

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+import android.text.Html
+import android.util.Log
+import android.widget.TextView
+import com.bumptech.glide.Glide
+import com.bumptech.glide.request.RequestOptions
+import com.bumptech.glide.request.target.SimpleTarget
+import com.bumptech.glide.request.transition.Transition
+import java.net.URL
+
+//TODO (#169) :  Replace this with exploration asset downloader
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: `// TODO (#169) :  Replace this with exploration asset downloader`

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+/** UrlImage Parser for android TextView to extract image from Html content. */
+class UrlImageParser(internal var tvContents: TextView, internal var context: Context) : Html.ImageGetter {
+
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:25]: Remove extra blank lines

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+  val GCS_PREFIX: String = "https://storage.googleapis.com/"
+  val GCS_RESOURCE_BUCKET_NAME = "oppiaserver-resources/"
+  var  IMAGE_DOWNLOAD_URL_TEMPLATE = "/<entity_type>/<entity_id>/assets/image/<filename>"
+
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:29]: Remove extra blank lines

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+  var targets: ArrayList<BitmapTarget>? = null
+
+  override fun getDrawable(url: String): Drawable {
+    IMAGE_DOWNLOAD_URL_TEMPLATE = "exploration/umPkwp0L1M0-/assets/image/"
+
+    Log.d("url","htmlContent: " + url)
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: Remove this Log and also its `import`

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+    targets?.add(target)
+    load.into(target)
+    return urlDrawable
+  }
+
+  inner class BitmapTarget(private val urlDrawable: UrlDrawable) : SimpleTarget<Bitmap>() {
+    internal var drawable: Drawable? = null
+
+    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: Remove extra blank lines

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+      drawable = BitmapDrawable(context.resources, resource)
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: Remove extra blank lines

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+        val ww = (drawable as BitmapDrawable).intrinsicWidth
+        val newHeight = hh * w / ww
+        val rect = Rect(0, 0, w, newHeight)
+        (drawable as BitmapDrawable).bounds = rect
+        urlDrawable.bounds = rect
+        urlDrawable.drawable = drawable
+        tvContents.text = tvContents.text
+        tvContents.invalidate()
+      }
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: Remove extra blank lines

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  inner class UrlDrawable : BitmapDrawable() {
+    var drawable: Drawable? = null
+    override fun draw(canvas: Canvas) {
+      if (drawable != null)
+        drawable!!.draw(canvas)
+    }
+  }
+
```
Comment (rt4914) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: Remove extra blank lines

**Review by rt4914 — COMMENTED** [review]
@veena14cs Thanks.
I have taken an overview review of this PR, finish these and later on I can take in-depth look at this PR.

**Comment by rt4914** [comment]
@veena14cs Please add test-cases for this PR

File: app/src/main/java/org/oppia/app/fragment/FragmentComponent.kt (hunk 1 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/fragment/FragmentComponent.kt]: corrected variable name

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 2 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: updated comment added missing punctuation and space

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 3 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: To differentiate I made it Learner. But `VIEW_TYPE_INTERACTION` sounds good. Will change it.

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 4 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: removed blank line

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 5 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: made this condition on multiple lines.

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt (hunk 24 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: removed log

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt (hunk 27 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: removed extra spaces

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt (hunk 28 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: removed blank lines

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt (hunk 25 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: updated this file

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt (hunk 26 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: removed spaces

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 6 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: declared const val on top

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 7 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: removed spaces

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 8 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: removed spaces

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 9 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: removed spaces

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt (hunk 10 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt:23]: removed extra lines

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt (hunk 11 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt]: done

File: app/src/main/res/drawable/bg_blue_card.xml (hunk 12 above)
Comment (veena14cs) [reply app/src/main/res/drawable/bg_blue_card.xml]: It looks too thick if I update to 2.

File: app/src/main/res/drawable/bg_blue_card.xml (hunk 13 above)
Comment (veena14cs) [reply app/src/main/res/drawable/bg_blue_card.xml]: updated to 8dp

File: app/src/main/res/drawable/bg_white_card.xml (hunk 14 above)
Comment (veena14cs) [reply app/src/main/res/drawable/bg_white_card.xml]: corrected

File: app/src/main/res/layout/content_card_items.xml (hunk 16 above)
Comment (veena14cs) [reply app/src/main/res/layout/content_card_items.xml]: corrected

File: app/src/main/res/layout/content_card_items.xml (hunk 17 above)
Comment (veena14cs) [reply app/src/main/res/layout/content_card_items.xml]: done

File: app/src/main/res/layout/content_list_fragment.xml (hunk 18 above)
Comment (veena14cs) [reply app/src/main/res/layout/content_list_fragment.xml]: corrected

File: app/src/main/res/layout/learners_card_item.xml (hunk 19 above)
Comment (veena14cs) [reply app/src/main/res/layout/learners_card_item.xml]: corrected

File: app/src/main/res/layout/learners_card_item.xml (hunk 20 above)
Comment (veena14cs) [reply app/src/main/res/layout/learners_card_item.xml]: updated to 24dp

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt (hunk 21 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/UrlImageParser.kt]: Done

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt (hunk 22 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:25]: removed extra spaces from everywhere

File: app/src/main/res/drawable/bg_white_card.xml (hunk 15 above)
Comment (veena14cs) [reply app/src/main/res/drawable/bg_white_card.xml]: changed to 2dp

**Review by rt4914 — COMMENTED** [review]
LGTM

**Comment by rt4914** [comment]
@BenHenning PTAL final review

File: app/build.gradle [hunk 29]
```diff
@@ -64,7 +64,8 @@ dependencies {
       'com.google.dagger:dagger:2.24',
       "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version",
       "org.jetbrains.kotlinx:kotlinx-coroutines-core:1.2.1",
-      "org.jetbrains.kotlinx:kotlinx-coroutines-android:1.2.1"
+      "org.jetbrains.kotlinx:kotlinx-coroutines-android:1.2.1",
```
Comment (BenHenning) [inline app/build.gradle]: Super nit: please use single quote strings unless you're including a variable. Probably ditto for the kotlinx-coroutines-core bit above.

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 30]
```diff
@@ -0,0 +1,110 @@
...
+const val HTML_TAG = "img"
+const val CUSTOM_ATTRIBUTE = "filepath-with-value"
+const val HTML_ATTRIBUTE = "src"
+
+/** Adapter to bind the contents to the [RecyclerView]. It handles rich-text content. */
+class ContentCardAdapter(
+  private val context: Context,
+  private val entity_type: String,
+  private val entity_id: String,
+  val contentList: MutableList<GaeSubtitledHtml>
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Please use the proto SubtitledHtml instead of GAE.

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 31]
```diff
@@ -0,0 +1,110 @@
...
+  inner class ContentViewHolder(val binding: ViewDataBinding) : RecyclerView.ViewHolder(binding.root) {
+    fun bind(rawString: String?) {
+      binding.setVariable(BR.htmlContent, rawString)
+      binding.executePendingBindings();
+      val html: Spannable = parseHtml(rawString, binding.root.tvContents)
+      binding.root.tvContents.text = html
+    }
+  }
+
+  inner class LearnersViewHolder(val binding: ViewDataBinding) : RecyclerView.ViewHolder(binding.root) {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Should this maybe be ``InteractionViewHolder`` for parity with the type above & the corresponding ``ContentViewHolder``?

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 32]
```diff
@@ -0,0 +1,110 @@
...
+      VIEW_TYPE_CONTENT
+    }
+  }
+
+  override fun getItemCount(): Int {
+    return contentList!!.size
+  }
+
+  inner class ContentViewHolder(val binding: ViewDataBinding) : RecyclerView.ViewHolder(binding.root) {
+    fun bind(rawString: String?) {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Nit here & below: please make this method internal since it shouldn't be public.

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 33]
```diff
@@ -0,0 +1,110 @@
...
+    } else {
+      VIEW_TYPE_CONTENT
+    }
+  }
+
+  override fun getItemCount(): Int {
+    return contentList!!.size
+  }
+
+  inner class ContentViewHolder(val binding: ViewDataBinding) : RecyclerView.ViewHolder(binding.root) {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Nit here & below: please make this private since this class doesn't need to be exposed outside of the adapter.

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 34]
```diff
@@ -0,0 +1,110 @@
...
+  inner class LearnersViewHolder(val binding: ViewDataBinding) : RecyclerView.ViewHolder(binding.root) {
+    fun bind(rawString: String?) {
+      binding.setVariable(BR.htmlContent, rawString)
+      binding.executePendingBindings();
+      val html: Spannable = parseHtml(rawString, binding.root.tvContents)
+      binding.root.tvContents.text = html
+    }
+  }
+
+  private fun parseHtml(rawString: String?, tvContents: TextView): Spannable {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Similar to my comment in your other PR--let's move this to a general utility so that we don't need to repeat it.

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt [hunk 35]
```diff
@@ -0,0 +1,110 @@
...
+import org.oppia.data.backends.gae.model.GaeSubtitledHtml
+import org.oppia.util.data.UrlImageParser
+
+const val CUSTOM_TAG = "oppia-noninteractive-image"
+const val HTML_TAG = "img"
+const val CUSTOM_ATTRIBUTE = "filepath-with-value"
+const val HTML_ATTRIBUTE = "src"
+
+/** Adapter to bind the contents to the [RecyclerView]. It handles rich-text content. */
+class ContentCardAdapter(
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt:20]: Generally looks good, though I suggest at least adding a TODO to move this over to #172 once that solution is available.

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt [hunk 36]
```diff
@@ -0,0 +1,62 @@
...
+    val binding = ContentListFragmentBinding.inflate(inflater, container, /* attachToRoot= */ false)
+    binding.recyclerView.apply {
+      binding.recyclerView.layoutManager = LinearLayoutManager(context)
+      binding.contentCardAdapter = ContentCardAdapter(context, entity_type, entity_id, contentList);
+    }
+    fetchDummyExplorations()
+    return binding.root
+  }
+
+  // TODO (#121) :Replace this once interface for ExplorationDataController is available
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt]: Nit: please format these like:

``TODO(#121): Replace this once...``

(Note the spacing for symbols).

File: app/src/main/res/layout/content_list_fragment.xml [hunk 37]
```diff
@@ -0,0 +1,30 @@
...
+    tools:context=".player.content.ContentListFragment">
+    <androidx.recyclerview.widget.RecyclerView
+      android:id="@+id/recyclerView"
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:layout_alignParentTop="true"
+      android:layout_marginStart="8dp"
+      android:layout_marginTop="4dp"
+      android:layout_marginEnd="8dp"
+      android:adapter="@{contentCardAdapter}"
```
Comment (BenHenning) [inline app/src/main/res/layout/content_list_fragment.xml]: Neat. I didn't know you could do this.

Don't we still need to notify when the adapter changes?

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt [hunk 38]
```diff
@@ -0,0 +1,73 @@
+package org.oppia.util.data
```
Comment (BenHenning) [inline utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:1]: Repeat from an earlier PR: please send out a separate PR to submit this with its own tests. I'd like to review it outside the context of these other PRs, though it does mean you need to base these PRs on the branch introducing UrlImageParser.

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt [hunk 39]
```diff
@@ -0,0 +1,62 @@
...
+  private fun fetchDummyExplorations() {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt]: Can we fetch this from ``ExplorationDataController``, instead? Let's go ahead and change those explorations to contain any additional data we need to properly test this code (feel free to copy the html samples you have below into those explorations if needed).

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt [hunk 40]
```diff
@@ -0,0 +1,62 @@
...
+import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+import androidx.recyclerview.widget.LinearLayoutManager
+import org.oppia.app.application.ApplicationContext
+import org.oppia.app.databinding.ContentListFragmentBinding
+import org.oppia.data.backends.gae.model.GaeSubtitledHtml
+import javax.inject.Inject
+
+/** Presenter for [ContentListFragment]. */
+class ContentListFragmentPresenter @Inject constructor(
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt:16]: Please add tests for ``ContentListFragment``.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @veena14cs! This PR looks really solid, however I'd really like us to submit the image parser separately since I want to spend a bit more time looking at it in isolation, and seeing tests for it.

At a high-level, we should also be adding tests for all new UI components being introduced, we should ensure that we're using domain models instead of GAE models, and we should hook into domain controllers rather than faking out data in the UI (only fake data if there isn't a controller available, and please let me know right away if that happens since the infrastructure team is actively working to keep UI structure work unblocked).

Please let me know if you have any concerns or doubts with any of these. My intent is to ensure that there's as little work as possible beyond layouts and animations when finalizing these UIs later on. Ie, we should try to get the code structure itself as close to final as possible with these PRs.

**Review by nikitamarysolomanpvt — APPROVED** [review]
LGTM @veena14cs Please add test-cases for this PR

File: app/build.gradle (hunk 29 above)
Comment (veena14cs) [reply app/build.gradle]: done

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 31 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: Corrected Viewholder name as per parity

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 33 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: made inner class private

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt (hunk 36 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt]: corrected todo format

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt (hunk 39 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt]: Having problem in fetching explorations. Its returning null.

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt (hunk 40 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt]: done

Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt:16]: added test cases for contentlistFragment

File: app/src/main/res/layout/content_list_fragment.xml (hunk 37 above)
Comment (veena14cs) [reply app/src/main/res/layout/content_list_fragment.xml]: removed setting of adapter from xml and added in the Fragment

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 32 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: made it internal

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 35 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt:20]: added TODO to adapter

File: utility/src/main/java/org/oppia/util/data/UrlImageParser.kt (hunk 38 above)
Comment (veena14cs) [reply utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:1]: Separate PR [#201 ](https://github.com/oppia/oppia-android/pull/201) is been made for UrlImageParser.

File: app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt (hunk 39 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentListFragmentPresenter.kt]: now fetching data from exploration

File: app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt (hunk 30 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/app/player/content/ContentCardAdapter.kt]: used data from ContentViewModel.

**Comment by veena14cs** [comment]
Duplicate of this PR is created [here](https://github.com/oppia/oppia-android/pull/228).
